DynamicTablesPkg/Library/Acpi/Common/AcpiSlitLib/SlitGenerator.c

```diff
@@ -0,0 +1,412 @@
+      DEBUG ((
+        DEBUG_ERROR,
+        "ERROR: SLIT: Invalid relation value %d\n",
+        DomainRelationInfo[Index].Relation
+        ));
+      return EFI_INVALID_PARAMETER;
+    }
+
+    Status = GetEArchCommonObjProxDomainInfo (
+               CfgMgrProtocol,
+               DomainRelationInfo[Index].FirstDomainToken,
+               &DomainInfoFirst,
+               NULL
+               );
+    if (EFI_ERROR (Status)) {
+      DEBUG ((
+        DEBUG_ERROR,
+        "ERROR: SLIT: Failed to retrieve First Proximity Domain information. Status = %r\n",
+        Status
+        ));
+      return Status;
+    }
+
+    Status = GetEArchCommonObjProxDomainInfo (
+               CfgMgrProtocol,
+               DomainRelationInfo[Index].SecondDomainToken,
+               &DomainInfoSecond,
+               NULL
+               );
+    if (EFI_ERROR (Status)) {
+      DEBUG ((
+        DEBUG_ERROR,
+        "ERROR: SLIT: Failed to retrieve Second Proximity Domain information. Status = %r\n",
+        Status
+        ));
+      return Status;
+    }
+
+    if ((DomainInfoFirst == NULL) || (DomainInfoSecond == NULL)) {
+      DEBUG ((
+        DEBUG_ERROR,
+        "ERROR: SLIT: No Proximity Domain information provided by configuration manager.\n"
+        ));
+      return EFI_INVALID_PARAMETER;
+    }
+
+    if (DomainInfoFirst->DomainId == DomainInfoSecond->DomainId) {
+      if (DomainRelationInfo[Index].Relation != 0x10) {
+        DEBUG ((
+          DEBUG_ERROR,
+          "ERROR: SLIT: Invalid relation value %d for same domain ID %d\n",
+          DomainRelationInfo[Index].Relation,
+          DomainInfoFirst->DomainId
+          ));
+        return EFI_INVALID_PARAMETER;
+      }
+    }
+
+    NumberOfSystemLocalities = MAX (
+                                 NumberOfSystemLocalities,
+                                 MAX (
+                                   DomainInfoFirst->DomainId,
+                                   DomainInfoSecond->DomainId
+                                   )
+                                 );
+  }
+
+  /// Since locality IDs start from 0, increment the count by 1
+  NumberOfSystemLocalities++;
+
+  AcpiSlitTable = AllocateZeroPool (
+                    sizeof (EFI_ACPI_6_5_SYSTEM_LOCALITY_DISTANCE_INFORMATION_TABLE_HEADER) +
+                    (sizeof (UINT8) * NumberOfSystemLocalities * NumberOfSystemLocalities)
+                    );
+  if (AcpiSlitTable == NULL) {
+    DEBUG (
+      (DEBUG_ERROR, "ERROR: SLIT: Failed to allocate memory for SLIT table.\n"));
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  SlitEntry = (UINT8 *)(AcpiSlitTable + sizeof (EFI_ACPI_6_5_SYSTEM_LOCALITY_DISTANCE_INFORMATION_TABLE_HEADER));
+
+  for (Index = 0; Index < NumberOfSystemLocalities * NumberOfSystemLocalities; Index++) {
+    /// Set the non-diagonal elements to 0xFF
+    SlitEntry[Index] = 0xFF;
+    /// Set the diagonal elements to 0x10
+    if ((Index % (NumberOfSystemLocalities + 1)) == 0) {

```

Comment:
Isn't it populating the first column of the array ?
I think it might be simpler to use:
```
SetMem (SlitEntry, NumberOfSystemLocalities * NumberOfSystemLocalities, 0xFF);

for (Index = 0; Index < NumberOfSystemLocalities; Index++) {
  SlitEntry[Index * Index] = 0x10;
}
```

Reply:
The Index computation I wrote is wrong: it should be ``` Index * NumberOfSystemLocalities + Index ```